File: javascript/src/resources/analysis.ts
/** AI analysis endpoints and the product documentation assistant. */

import type { Transport } from '../transport.js';
import type { JsonObject } from '../types.js';
import type { CallOptions } from './companies.js';

const FILING_EXTRACT_MODES = ['extract', 'verify'] as const;
export type FilingExtractMode = (typeof FILING_EXTRACT_MODES)[number];

export interface AnalyseCompanyOptions extends CallOptions {
  /** Rule set whose findings should frame the analysis. */
  ruleSetId?: string;
  /** Override the AI provider, e.g. `"claude"`, `"gpt"`, `"gemini"`. */
  provider?: string;
  /** Restrict the write-up to named sections. At most 50. */
  sections?: string[];
  /** FCA Firm Reference Number, when the company is regulated and you know it. */
  fcaFrn?: string;
  /**
   * Evidence gathered elsewhere — confirmed screening hits, selected filings,
   * analyst-confirmed identity links — folded into the prompt so the analysis
   * reflects your review rather than raw unconfirmed matches.
   */
  context?: JsonObject;
}

export interface AdverseMediaOverviewOptions extends CallOptions {
  provider?: string;
  /** Search results to summarise, as returned by `news.screenCompany()`. */
  results?: JsonObject;
}

export interface FilingExtractOptions extends CallOptions {
  /** `"extract"` pulls structured data out; `"verify"` checks an existing extraction. */
  mode?: FilingExtractMode;
  provider?: string;
}

/** One prior turn of a documentation-assistant conversation. */
export interface DocsTurn {
  role: 'user' | 'assistant';
  content: string;
}

export interface DocsAskOptions extends CallOptions {
  /** Prior turns for follow-up questions. At most 12. */
  history?: DocsTurn[];
}

/**
 * LLM-written analyst commentary over an assessment.
 *
 * **Requires an active Professional subscription**, except for {@link status}.
 * These calls invoke a language model, so they are slower and dearer than the
 * deterministic rule engine — and, like any LLM output, the narrative is a
 * drafting aid for an analyst, not a compliance decision.
 */
export class Analysis {
  constructor(private readonly transport: Transport) {}

  /** Which AI providers are configured. Available without a subscription. */
  async status(options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/analysis/status', options);
  }

  /** Write a narrative risk analysis for one company. */
  async company(companyNumber: string, options: AnalyseCompanyOptions = {}): Promise<JsonObject> {
    const body: JsonObject = { company_number: companyNumber };
    if (options.ruleSetId !== undefined) body.rule_set_id = options.ruleSetId;
    if (options.provider !== undefined) body.provider = options.provider;
    if (options.sections?.length) {
      if (options.sections.length > 50) {
        throw new TypeError(`sections accepts at most 50 entries, got ${options.sections.length}`);
      }
      body.sections = options.sections;
    }
    if (options.fcaFrn !== undefined) body.fca_frn = options.fcaFrn;
    if (options.context !== undefined) body.context = options.context;

    return this.transport.post('/analysis/company', { body, signal: options.signal });
  }

  /** Summarise adverse media results into a short overview. */
  async adverseMediaOverview(
    companyNumber: string,
    options: AdverseMediaOverviewOptions = {},
  ): Promise<JsonObject> {
    const body: JsonObject = { company_number: companyNumber };
    if (options.provider !== undefined) body.provider = options.provider;
    if (options.results !== undefined) body.results = options.results;

    return this.transport.post('/analysis/adverse-media-overview', {
      body,
      signal: options.signal,
    });
  }

  /** Read a filing PDF with an LLM. */
  async filingExtract(
    companyNumber: string,
    transactionId: string,
    options: FilingExtractOptions = {},
  ): Promise<JsonObject> {
    const mode = options.mode ?? 'extract';
    if (!FILING_EXTRACT_MODES.includes(mode)) {
      throw new TypeError(
        `mode must be one of ${FILING_EXTRACT_MODES.join(', ')}, got '${String(mode)}'`,
      );
    }

    const body: JsonObject = {
      company_number: companyNumber,
      transaction_id: transactionId,
      mode,
    };
    if (options.provider !== undefined) body.provider = options.provider;

    return this.transport.post('/analysis/filing-extract', { body, signal: options.signal });
  }
}

/** Ask questions about the product in natural language. */
export class Docs {
  constructor(private readonly transport: Transport) {}

  /** Ask the documentation assistant a question. Up to 2,000 characters. */
  async ask(message: string, options: DocsAskOptions = {}): Promise<JsonObject> {
    if (!message || !message.trim()) {
      throw new TypeError('message must be a non-empty string');
    }
    const body: JsonObject = { message };
    if (options.history?.length) {
      if (options.history.length > 12) {
        throw new TypeError(`history accepts at most 12 turns, got ${options.history.length}`);
      }
      body.history = options.history;
    }
    return this.transport.post('/docs/ask', { body, signal: options.signal });
  }
}

File: elixir/lib/kyccentral/analysis.ex
defmodule KYCCentral.Analysis do
  @moduledoc """
  LLM-written analyst commentary over an assessment.

  **Requires an active Professional subscription**, except for `status/1`. These
  calls invoke a language model, so they are slower and dearer than the
  deterministic rule engine — and, like any LLM output, the narrative is a
  drafting aid for an analyst, not a compliance decision.
  """

  alias KYCCentral.{Error, Transport}

  @type result :: {:ok, map()} | {:error, Error.t()}

  @filing_extract_modes ~w(extract verify)
  @max_sections 50

  @doc "Which AI providers are configured. Available without a subscription."
  @spec status(KYCCentral.t()) :: result()
  def status(client), do: Transport.request(client, :get, "/analysis/status")

  @doc """
  Write a narrative risk analysis for one company.

  ## Options

    * `:rule_set_id` — rule set whose findings should frame the analysis.
    * `:provider` — override the AI provider, e.g. `"claude"`, `"gpt"`,
      `"gemini"`. Defaults to whichever the API is configured with.
    * `:sections` — restrict the write-up to named sections. At most
      #{@max_sections}.
    * `:fca_frn` — FCA Firm Reference Number, when the company is regulated and
      you already know it.
    * `:context` — evidence gathered elsewhere (confirmed screening hits,
      selected filings, analyst-confirmed identity links) folded into the prompt,
      so the analysis reflects your review rather than raw unconfirmed matches.
  """
  @spec company(KYCCentral.t(), String.t(), keyword()) :: result()
  def company(client, company_number, opts \\ []) do
    with {:ok, sections} <- validate_sections(opts[:sections]) do
      body =
        %{"company_number" => company_number}
        |> put_optional("rule_set_id", opts[:rule_set_id])
        |> put_optional("provider", opts[:provider])
        |> put_optional("sections", sections)
        |> put_optional("fca_frn", opts[:fca_frn])
        |> put_optional("context", opts[:context])

      Transport.request(client, :post, "/analysis/company", body: body)
    end
  end

  @doc """
  Summarise adverse media results into a short overview.

  ## Options

    * `:provider`
    * `:results` — search results to summarise, as returned by
      `KYCCentral.News.screen_company/2` or `KYCCentral.News.search_names/2`.
  """
  @spec adverse_media_overview(KYCCentral.t(), String.t(), keyword()) :: result()
  def adverse_media_overview(client, company_number, opts \\ []) do
    body =
      %{"company_number" => company_number}
      |> put_optional("provider", opts[:provider])
      |> put_optional("results", opts[:results])

    Transport.request(client, :post, "/analysis/adverse-media-overview", body: body)
  end

  @doc """
  Read a filing PDF with an LLM.

  ## Options

    * `:mode` — `"extract"` to pull structured data out of the filing, or
      `"verify"` to check an existing extraction against the document. Defaults
      to `"extract"`.
    * `:provider`
  """
  @spec filing_extract(KYCCentral.t(), String.t(), String.t(), keyword()) :: result()
  def filing_extract(client, company_number, transaction_id, opts \\ []) do
    mode = Keyword.get(opts, :mode, "extract")

    if mode in @filing_extract_modes do
      body =
        %{
          "company_number" => company_number,
          "transaction_id" => transaction_id,
          "mode" => mode
        }
        |> put_optional("provider", opts[:provider])

      Transport.request(client, :post, "/analysis/filing-extract", body: body)
    else
      {:error,
       Error.invalid_argument(
         ":mode must be one of #{Enum.join(@filing_extract_modes, ", ")}, got #{inspect(mode)}"
       )}
    end
  end

  defp validate_sections(nil), do: {:ok, nil}
  defp validate_sections([]), do: {:ok, nil}

  defp validate_sections(sections) when is_list(sections) do
    if length(sections) > @max_sections do
      {:error,
       Error.invalid_argument(
         ":sections accepts at most #{@max_sections} entries, got #{length(sections)}"
       )}
    else
      {:ok, sections}
    end
  end

  defp validate_sections(other) do
    {:error, Error.invalid_argument(":sections must be a list, got #{inspect(other)}")}
  end

  defp put_optional(body, _key, nil), do: body
  defp put_optional(body, key, value), do: Map.put(body, key, value)
end

defmodule KYCCentral.Docs do
  @moduledoc "Ask questions about the product in natural language."

  alias KYCCentral.{Error, Transport}

  @max_history 12

  @doc """
  Ask the documentation assistant a question.

  ## Options

    * `:history` — prior turns for follow-up questions, at most #{@max_history},
      each `%{role: "user" | "assistant", content: "..."}`.
  """
  @spec ask(KYCCentral.t(), String.t(), keyword()) :: {:ok, map()} | {:error, Error.t()}
  def ask(client, message, opts \\ []) do
    history = opts[:history] || []

    cond do
      not is_binary(message) or String.trim(message) == "" ->
        {:error, Error.invalid_argument("message must be a non-empty string")}

      length(history) > @max_history ->
        {:error,
         Error.invalid_argument(
           ":history accepts at most #{@max_history} turns, got #{length(history)}"
         )}

      true ->
        body = %{"message" => message}
        body = if history == [], do: body, else: Map.put(body, "history", history)
        Transport.request(client, :post, "/docs/ask", body: body)
    end
  end
end
